File: combo2-advanced-agentic-engineering/expense-categoriser-csharp/tests/ExpenseCategoriser.Tests/ApiTests.cs
using System.Net;
using System.Net.Http.Json;
using System.Text.Json;

using ExpenseCategoriser;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.Extensions.DependencyInjection;
using Xunit;

namespace ExpenseCategoriser.Tests;

/// <summary>
/// Layer 2 — API contract tests. Boots the real app in-process with a fake LLM.
/// </summary>
/// <remarks>
/// These check the shape of the HTTP contract, not the quality of the answer:
/// status codes, response fields, and — most importantly — that a contract
/// violation from the model becomes a 502 rather than a 500 or a crash.
/// Still no key, still free.
/// </remarks>
public class ApiTests
{
    private static WebApplicationFactory<Program> AppWith(ILlmClient client) =>
        new WebApplicationFactory<Program>().WithWebHostBuilder(builder =>
        {
            builder.ConfigureServices(services =>
            {
                services.AddSingleton(client);
            });
        });

    [Fact]
    public async Task Health_ReturnsOk()
    {
        using var app = AppWith(FakeLlmClient.Returning("Other", 0.9));
        using var http = app.CreateClient();

        var response = await http.GetAsync("/health", TestContext.Current.CancellationToken);

        Assert.Equal(HttpStatusCode.OK, response.StatusCode);
    }

    [Fact]
    public async Task Categories_ReturnsTheCanonicalList()
    {
        using var app = AppWith(FakeLlmClient.Returning("Other", 0.9));
        using var http = app.CreateClient();

        var payload = await http.GetFromJsonAsync<JsonElement>(
            "/categories", TestContext.Current.CancellationToken);

        var categories = payload.GetProperty("categories")
            .EnumerateArray().Select(e => e.GetString()).ToArray();

        Assert.Equal(Categories.Canonical, categories);
    }

    [Fact]
    public async Task Categorise_ReturnsTheModelsAnswer()
    {
        using var app = AppWith(FakeLlmClient.Returning("Food & Dining", 0.95));
        using var http = app.CreateClient();

        var response = await http.PostAsJsonAsync(
            "/categorise",
            new { description = "Starbucks Coffee", amount = 5.45 },
            TestContext.Current.CancellationToken);

        Assert.Equal(HttpStatusCode.OK, response.StatusCode);

        var body = await response.Content.ReadFromJsonAsync<CategorisationOut>(
            TestContext.Current.CancellationToken);

        Assert.Equal("Food & Dining", body!.Category);
        Assert.False(body.UsedFallback);
    }

    [Fact]
    public async Task Categorise_LowConfidenceIsStillA200()
    {
        // The fallback is a successful response. The client needs to know the
        // model wasn't confident — not that the request failed.
        using var app = AppWith(FakeLlmClient.Returning("Travel", 0.2));
        using var http = app.CreateClient();

        var response = await http.PostAsJsonAsync(
            "/categorise",
            new { description = "Something ambiguous", amount = 12.0 },
            TestContext.Current.CancellationToken);

        Assert.Equal(HttpStatusCode.OK, response.StatusCode);

        var body = await response.Content.ReadFromJsonAsync<CategorisationOut>(
            TestContext.Current.CancellationToken);

        Assert.Equal("Other", body!.Category);
        Assert.True(body.UsedFallback);
    }

    [Fact]
    public async Task Categorise_ContractViolationBecomes502()
    {
        // Not a 500. The service is healthy; the model misbehaved. This
        // distinction is what lets the CE pipeline alert on model drift without
        // drowning in ordinary server errors.
        using var app = AppWith(new FakeLlmClient("this is not json"));
        using var http = app.CreateClient();

        var response = await http.PostAsJsonAsync(
            "/categorise",
            new { description = "Starbucks", amount = 5.45 },
            TestContext.Current.CancellationToken);

        Assert.Equal(HttpStatusCode.BadGateway, response.StatusCode);
    }

    [Fact]
    public async Task Categorise_MissingApiKeyBecomes500()
    {
        using var app = AppWith(new ThrowingLlmClient(
            new InvalidOperationException("GOOGLE_API_KEY is not set.")));
        using var http = app.CreateClient();

        var response = await http.PostAsJsonAsync(
            "/categorise",
            new { description = "Starbucks", amount = 5.45 },
            TestContext.Current.CancellationToken);

        Assert.Equal(HttpStatusCode.InternalServerError, response.StatusCode);
    }

    [Fact]
    public async Task Categorise_RejectsEmptyDescription()
    {
        using var app = AppWith(FakeLlmClient.Returning("Other", 0.9));
        using var http = app.CreateClient();

        var response = await http.PostAsJsonAsync(
            "/categorise",
            new { description = "", amount = 5.45 },
            TestContext.Current.CancellationToken);

        Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
    }
}

File: combo2-advanced-agentic-engineering/java/expense-categoriser-java/src/test/java/ai/octoco/expensecategoriser/ApiTests.java
package ai.octoco.expensecategoriser;

import static org.assertj.core.api.Assertions.assertThat;

import ai.octoco.expensecategoriser.Models.CategorisationOut;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;
import org.springframework.context.annotation.Primary;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

/**
 * Layer 2 — API contract tests. Boots the real app with a fake LLM.
 *
 * <p>These check the shape of the HTTP contract, not the quality of the
 * answer: status codes, response fields, and — most importantly — that a
 * contract violation from the model becomes a 502 rather than a 500 or a
 * crash. Still no key, still free.
 *
 * <p>One Spring context is shared across every test in this class — see
 * {@link DelegatingLlmClient} for why, and how each test points it at the
 * fake it needs.
 */
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@Import(ApiTests.TestConfig.class)
class ApiTests {

    @TestConfiguration
    static class TestConfig {
        @Bean
        @Primary
        DelegatingLlmClient testLlmClient() {
            return new DelegatingLlmClient();
        }
    }

    @Autowired
    private TestRestTemplate rest;

    @Autowired
    private DelegatingLlmClient llmClient;

    @Test
    void health_returnsOk() {
        ResponseEntity<Map> response = rest.getForEntity("/health", Map.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
    }

    @Test
    void categories_returnsTheCanonicalList() {
        ResponseEntity<Map> response = rest.getForEntity("/categories", Map.class);

        @SuppressWarnings("unchecked")
        List<String> categories = (List<String>) response.getBody().get("categories");
        assertThat(categories).containsExactlyElementsOf(Models.CANONICAL_CATEGORIES);
    }

    @Test
    void categorise_returnsTheModelsAnswer() {
        llmClient.setDelegate(FakeLlmClient.returning("Food & Dining", 0.95));

        ResponseEntity<CategorisationOut> response = rest.postForEntity(
                "/categorise",
                Map.of("description", "Starbucks Coffee", "amount", 5.45),
                CategorisationOut.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody().category()).isEqualTo("Food & Dining");
        assertThat(response.getBody().usedFallback()).isFalse();
    }

    @Test
    void categorise_lowConfidenceIsStillA200() {
        // The fallback is a successful response. The client needs to know the
        // model wasn't confident — not that the request failed.
        llmClient.setDelegate(FakeLlmClient.returning("Travel", 0.2));

        ResponseEntity<CategorisationOut> response = rest.postForEntity(
                "/categorise",
                Map.of("description", "Something ambiguous", "amount", 12.0),
                CategorisationOut.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody().category()).isEqualTo("Other");
        assertThat(response.getBody().usedFallback()).isTrue();
    }

    @Test
    void categorise_contractViolationBecomes502() {
        // Not a 500. The service is healthy; the model misbehaved. This
        // distinction is what lets the CE pipeline alert on model drift
        // without drowning in ordinary server errors.
        llmClient.setDelegate(new FakeLlmClient("this is not json"));

        ResponseEntity<String> response = rest.postForEntity(
                "/categorise",
                Map.of("description", "Starbucks", "amount", 5.45),
                String.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_GATEWAY);
    }

    @Test
    void categorise_missingApiKeyBecomes500() {
        llmClient.setDelegate(new ThrowingLlmClient(
                new IllegalStateException("GOOGLE_API_KEY is not set.")));

        ResponseEntity<String> response = rest.postForEntity(
                "/categorise",
                Map.of("description", "Starbucks", "amount", 5.45),
                String.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR);
    }

    @Test
    void categorise_rejectsEmptyDescription() {
        llmClient.setDelegate(FakeLlmClient.returning("Other", 0.9));

        ResponseEntity<String> response = rest.postForEntity(
                "/categorise",
                Map.of("description", "", "amount", 5.45),
                String.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
    }
}
